### PR TITLE
feat(instance-private-network): Create/Read private network

### PR DESCRIPTION
**Pull request by Monitob**

### CODE REVIEW

File: docs/resources/instance_server.md [hunk 1]
```diff
@@ -110,6 +110,24 @@ resource "scaleway_instance_server" "web" {
...
+```hcl
+resource scaleway_vpc_private_network pn01 {
+    name = "private_network_instance"
+}
+
+resource "scaleway_instance_server" "base" {
+  image = "ubuntu_focal"
+  type  = "DEV1-S"
+
+  tags = [ "private_network" ]
```
Comment (remyleone) [inline docs/resources/instance_server.md:124]:
```suggestion
```

File: docs/resources/instance_server.md [hunk 2]
```diff
@@ -176,6 +197,17 @@ attached to the server. Updates to this field will trigger a stop/start of the s
 - `project_id` - (Defaults to [provider](../index.md#project_id) `project_id`) The ID of the project the server is associated with.
 
 
+## Private Network
+
+~> **Important:** Updates to `private_network` will recreate the attachment Instance.
```
Comment (remyleone) [inline docs/resources/instance_server.md:202]: What is an attachment Instance?

File: docs/resources/instance_server.md [hunk 3]
```diff
@@ -176,6 +197,17 @@ attached to the server. Updates to this field will trigger a stop/start of the s
...
+
+- `pn_id` - (Required) The private network ID where to connect.
+- `mac_address` The private NIC MAC address.
+- `status` The private NIC state.
+- `zone` - (Defaults to [provider](../index.md#zone) `zone`) The [zone](../guides/regions_and_zones.md#zones) in which the server should be created.
```
Comment (remyleone) [inline docs/resources/instance_server.md:207]: Maybe we should precise that private networks cannot be used across zones. You can only attach an instance in the same zone as a private network.

File: scaleway/helpers_instance.go [hunk 4]
```diff
@@ -242,3 +243,33 @@ func sanitizeVolumeMap(serverName string, volumes map[string]*instance.VolumeTem
...
+		if pnExist {
+			currentPN, err := vpcAPI.GetPrivateNetwork(&vpc.GetPrivateNetworkRequest{
+				PrivateNetworkID: expandID(privateNetworkID),
+				Zone:             server.Zone,
+			}, scw.WithContext(ctx))
+			if err != nil {
+				return nil, err
+			}
+			query := &instance.CreatePrivateNICRequest{
+				Zone: currentPN.Zone, ServerID: server.ID, PrivateNetworkID: currentPN.ID}
```
Comment (remyleone) [inline scaleway/helpers_instance.go:269]: Specify the fields in different lines to make it diff-friendly.

File: scaleway/resource_instance_server.go [hunk 5]
```diff
@@ -202,6 +203,34 @@ func resourceScalewayInstanceServer() *schema.Resource {
 					Type: schema.TypeString,
 				},
 			},
+			"private_network": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				MaxItems:    8,
```
Comment (remyleone) [inline scaleway/resource_instance_server.go:209]: Precise it in the documentation that we cannot have an instance in more than 8 different private networks.

File: scaleway/resource_instance_server.go [hunk 6]
```diff
@@ -702,6 +806,99 @@ func resourceScalewayInstanceServerUpdate(ctx context.Context, d *schema.Resourc
...
+	return nil
+}
+
+func (ph *PrivateNICsHandler) detach(o interface{}) error {
+	oPtr := expandStringPtr(o)
+	if oPtr != nil && len(*oPtr) > 0 {
+		// check if old private network still exist on instance server
+		if p, ok := ph.privateNICsMap[expandID(*oPtr)]; ok {
+			// detach private NIC
+			err := ph.instanceAPI.DeletePrivateNIC(&instance.DeletePrivateNICRequest{
```
Comment (remyleone) [inline scaleway/resource_instance_server.go:842]: Is it a synchronous operation? Or should a wait be added?

File: scaleway/resource_instance_server.go [hunk 7]
```diff
@@ -702,6 +806,99 @@ func resourceScalewayInstanceServerUpdate(ctx context.Context, d *schema.Resourc
...
+	return nil
+}
+
+func (ph *PrivateNICsHandler) attach(n interface{}) error {
+	nPtr := expandStringPtr(n)
+	if nPtr != nil {
+		// check if new private network was already attached on instance server
+		privateNetworkID := expandID(*nPtr)
+		if _, ok := ph.privateNICsMap[privateNetworkID]; !ok {
+			_, err := ph.instanceAPI.CreatePrivateNIC(&instance.CreatePrivateNICRequest{
```
Comment (remyleone) [inline scaleway/resource_instance_server.go:861]: Is it a synchronous operation? Or should a wait be added?

File: scaleway/resource_instance_server_test.go [hunk 8]
```diff
@@ -881,3 +882,133 @@ func TestAccScalewayInstanceServer_ServerWithBlockNonDefaultZone(t *testing.T) {
...
+					  image = "ubuntu_focal"
+					  type  = "DEV1-S"
+			
+				  	  tags = [ "private_network" ]
+					  private_network {
+						pn_id = scaleway_vpc_private_network.pn01.id
+					  }
+					}`,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckScalewayInstanceServerExists(tt, "scaleway_instance_server.base"),
```
Comment (remyleone) [inline scaleway/resource_instance_server_test.go:910]: Could you also check that the change happened on scaleway side too? With a `testAccCheckScalewayInstancePrivateNetworkAttachmentExists`

File: docs/resources/instance_server.md (hunk 2 above)
Comment (Monitob) [reply docs/resources/instance_server.md:202]: attachement is recreate a new interface (Private NIC)

File: scaleway/resource_instance_server.go (hunk 6 above)
Comment (Monitob) [reply scaleway/resource_instance_server.go:842]: It should be added, this check only if the private network was already detached by another process. For example by the console.

File: docs/resources/instance_server.md (hunk 2 above)
Comment (Monitob) [reply docs/resources/instance_server.md]: Please check the new comment

**Review by remyleone — APPROVED** [review]

**Pull request merged**